addons/format_lame.c

```diff
@@ -0,0 +1,522 @@
+		ast_log(LOG_WARNING, "LAME encoder returned error %d\n", res);
+		return -1;
+	} else if (res > 0) {
+		wres = fwrite(encoder_buffer, 1, res, fs->f);
+		if (wres != res) {
+			ast_log(LOG_WARNING, "Bad write (%d/%d): %s\n", wres, res, strerror(errno));
+			return -1;
+		} else {
+			ast_debug(3, "LAME encoder wrote %d bytes", wres);
+		}
+	}
+
+	return 0;
+}
+
+static off_t ast_lame_tell(struct ast_filestream *s)
+{
+	struct lame_private *p = s->_private;
+
+	if (p->decoder_counter != 0) {
+		return p->doffset;
+	} else {
+		return ftello(s->f) / BITS;
+	}
+}
+
+static int ast_lame_seek(struct ast_filestream *s, off_t sample_offset, int whence)
+{
+	struct lame_private *p = s->_private;
+
+	off_t offset = 0, min = 0, max, current;
+
+	sample_offset <<= 1;
+
+	if (p->decoder_counter != 0) {
+		max = p->total_bytes_decoded;
+		current = p->doffset;
+	} else {
+		if ((current = ftello(s->f)) < 0) {
+			ast_log(AST_LOG_WARNING, "Unable to determine current position in mp3 filestream %p: %s\n", s, strerror(errno));
+			return -1;
+		}
+
+		if (fseeko(s->f, 0, SEEK_END) < 0) {
+			ast_log(AST_LOG_WARNING, "Unable to seek to end of mp3 filestream %p: %s\n", s, strerror(errno));
+			return -1;
+		}
+
+		if ((max = ftello(s->f)) < 0) {
+			ast_log(AST_LOG_WARNING, "Unable to determine max position in mp3 filestream %p: %s\n", s, strerror(errno));
+			return -1;
+		}
+	}
+
+	if (whence == SEEK_SET)
+		offset = sample_offset;
+	else if (whence == SEEK_CUR || whence == SEEK_FORCECUR)
+		offset = sample_offset + current;
+	else if (whence == SEEK_END)
+		offset = max - sample_offset;
+	if (whence != SEEK_FORCECUR) {
+		offset = (offset > max)?max:offset;
+	}
+	/* always protect against seeking past begining. */
+	offset = (offset < min)?min:offset;
+
+	if (p->decoder_counter != 0) {
+		p->doffset = offset;
+		p->dsamples = (p->total_bytes_decoded - offset) / BITS;
+
+		return 0;
+	}
+
+	return fseeko(s->f, offset, SEEK_SET);
+}
+
+static int ast_lame_trunc(struct ast_filestream *s)
+{
+	int fd;
+	struct lame_private *p = s->_private;
+	off_t current = ast_lame_tell(s);
+
+	if (p->decoder_counter != 0) {
+
+		p->dsamples = 0;
+		p->total_bytes_decoded = p->doffset;
+
+		return 0;
+	} else {
+
+		if ((fd = fileno(s->f)) < 0) {
+			ast_log(AST_LOG_WARNING, "Unable to determine file descriptor for mp3 filestream %p: %s\n", s, strerror(errno));
+			return -1;
+		}
+
+		if ((current = ftello(s->f)) < 0) {
+			ast_log(AST_LOG_WARNING, "Unable to determine current position in mp3 filestream %p: %s\n", s, strerror(errno));
+			return -1;
+		}
+
+		return ftruncate(fd, current);
+	}
+}
+
+static char *ast_lame_getcomment(struct ast_filestream *s)
+{
+	char * comment = "Asterisk MP3 lame\0";
+	return comment;
+}
+
+static int ast_lame_write8(struct ast_filestream *fs, struct ast_frame *f){return ast_lame_write(fs, f, 8000);}
+static struct ast_frame *ast_lame_read8(struct ast_filestream *s, int *whennext){return ast_lame_read(s, whennext, SLIN_BUFLEN, 8000);}
```

Comment:
These should not be one-liners. If you want to create a macro because you feel this is repetitive that is fine.